[Android/iOS] Fix Shell.Background not working

jsuarezruiz opened the pull request:
### Description of Change ###

Implement `Shell.Background` property.

### Issues Resolved ### 

- fixes #13066 

### API Changes ###
 
 None

### Platforms Affected ### 

- iOS
- Android

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 
![fix13066](https://user-images.githubusercontent.com/6755973/101650346-1e8df280-3a3c-11eb-974f-a0c18e5eaaf9.gif)

### Testing Procedure ###

Launch Core Gallery and navigate to the issue 13066. If the NavBar background is a gradient, the test has passed.

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

jsuarezruiz committed:
Added repro sample

jsuarezruiz committed:
Added BackgroundProperty to Shell

jsuarezruiz committed:
Implement Shell BackgroundProperty on iOS

jsuarezruiz committed:
Updated sample

jsuarezruiz committed:
Updated appareance Background brush rotating the device

jsuarezruiz committed:
Implemented on Android

jsuarezruiz committed:
Merge branch '5.0.0' into fix-13066

jsuarezruiz committed:
Merge branch '5.0.0' into fix-13066

jsuarezruiz committed:
Fix build error

PureWeen commented on Xamarin.Forms.Platform.iOS/Renderers/IShellNavBarAppearanceTracker.cs:
```diff
@@ -6,6 +6,7 @@ namespace Xamarin.Forms.Platform.iOS
 
 	public interface IShellNavBarAppearanceTracker : IDisposable
 	{
+		ShellAppearance CurrentAppearance { get; }
```
It'd be nice to not have to modify this interface.  Can we just store the appearance locally to ShellSectionRenderer when `OnAppearanceChanged` is called?

jsuarezruiz replied:
Changed, see the last comment added.

PureWeen commented on Xamarin.Forms.Platform.Android/Renderers/ShellToolbarAppearanceTracker.cs:
```diff
@@ -18,18 +18,19 @@ public ShellToolbarAppearanceTracker(IShellContext shellContext)
 		}
 
-		protected virtual void SetColors(Toolbar toolbar, IShellToolbarTracker toolbarTracker, Color foreground, Color background, Color title)
+		protected virtual void SetColors(Toolbar toolbar, IShellToolbarTracker toolbarTracker, Color foreground, Color backgroundColor, Brush background, Color title)
```
We shouldn't modify this one because it's public

Can we create a brush version of this virtual?

protected virtual void SetColors(Toolbar toolbar, IShellToolbarTracker toolbarTracker, Brush foreground, Brush backgroundColor, Brush title)

Not sure if that will be too tricky to make non breaking from a behavior stand point but I think it works better then growing this method as we add brushes

jsuarezruiz replied:
Changed, see the last comment added.

PureWeen reviewed (changes requested):
This is a tricky one because there are other colors that default to BackgroundColor if they aren't set

https://github.com/xamarin/Xamarin.Forms/blob/72143a3650f04561488a037d840259a3deaca68c/Xamarin.Forms.Core/Shell/ShellAppearance.cs#L69

How hard would it be to make Brush versions of those?

jsuarezruiz committed:
Changes based on PR feedaback

jsuarezruiz commented:
@PureWeen Thanks for the feedback!
I have made some changes:
- Removed the `CurrentAppearance` property that I added to `IShellNavBarAppearanceTracker.cs`
- I have recover the original virtual method `SetColors` and added a  new `SetBrushes` method.
- I have tested everything on Android and iOS to verify that everything works as expected.

![fix13066](https://user-images.githubusercontent.com/6755973/105029618-defef100-5a52-11eb-9a6b-8bc3ad6ed801.gif)

Could you review the PR again?

bkaankose commented:
Are there any news on merging this one?

jfversluis commented:
Looks like a bigger change, let's not do this for Forms anymore at this point in time. Thanks for the time and effort everyone!
